include/mbedtls/ssl.h

```diff
@@ -1011,7 +1011,56 @@ typedef void mbedtls_ssl_export_keys_t( void *p_expkey,
  */
 struct mbedtls_ssl_config
 {
-    /* Group items by size (largest first) to minimize padding overhead */
+    /* Group items by size and reorder them to maximize usage of immediate offset access.    */
```

Comment:
I'm not convinced that size alone is the best order. Putting small fields first helps on arm thumb (which is our priority target when it comes to code size) because immediate offset access is of the form `address[offset * size]`, so the threshold for smaller item sizes is shorter than for larger item sizes. However, what matters in the end is that often-used items are under the threshold, and putting rarely-used items first can hurt that. Putting small items first can also make alignment suboptimal for efficient cache use on other platforms.

Reply:
Yeah, I know that.
Ordering by size will be not sufficient and the main goal here is to reorder structure fields based on which of them are more frequently used. 
BTW: can we say that we want to get benefits here from [locality of reference](https://en.wikipedia.org/wiki/Locality_of_reference) or this this term is only related to platform (processors) with cache?)  

After all, I decided to follow this idea: (https://github.com/ARMmbed/mbedtls/issues/4747#issuecomment-872713598) as a good first step. Now it's time to move further and take a look on those fields which are more commonly-used.

Reply:
Locality of reference means grouping together fields that are often used at the same time. That's for platforms with cache.

Reply:
Btw, here's a quick one-liner to get a rough idea of what fields of `ssl_context` are used most often:
```
sed -n 's/.*ssl->\([a-zA-Z0-9_]*\).*/\1/p' library/ssl_*.c | sort | uniq -c | sort -rn
```
It's far from perfect (for example, if multiple fields are used on the same line, it will only catch one of them) but could be a good start. Apparently the 10 most used fields would be something like:
```
    449 handshake
    360 conf
    134 session_negotiate
     94 out_msg
     93 state
     76 in_msg
     60 out_msglen
     49 session
     39 minor_ver
     39 in_hslen
```
Similar information for `ssl_config`, `ssl_session`, `ssl_handshake` and `ssl_transform` can be obtained by replacing `ssl` in the sed invication above with `conf`, `session[_a-z]*`, `handshake` and `transform[_a-z]*` respectively.

Reply:
Note: after discussing this with @daverodgman and @yanesca as well a re-reading the issue description which was about "quick wins" rather than "the best order", we think it's best to restrict the scope of the issue as far as 2.x is concerned. So for now, let's stick to what you've already done: reorder by field size. I've you've done further investigations already, please record them as comments in https://github.com/ARMmbed/mbedtls/issues/4747 but implementing further optimisations is now out of scope.

Also, since the primary reason for scheduling this work now was to get it in 2.x, we should focus on creating the 2.x backport as soon as possible (let's say, once this PR passes CI and gets one mostly-approving review).

Reply:
Backport for 2.x created: #4880. 
The only failure is now in API-ABI-checking but I guess it will always fail due to structure fields reorderin - Am I right?